src/python/pants/reporting/zipkin_reporter.py

```diff
@@ -90,45 +96,65 @@ def start_workunit(self, workunit):
           is_sampled=True,
         )
       else:
-        zipkin_attrs =  create_attrs_for_span(
+        zipkin_attrs = create_attrs_for_span(
           # trace_id is the same as run_uuid that is created in run_tracker and is the part of
           # pants_run id
           trace_id=self.trace_id,
           sample_rate=self.sample_rate, # Value between 0.0 and 100.0
         )
-        self.trace_id = zipkin_attrs.trace_id
         # TODO delete this line when parent_id will be passed in v2 engine:
         #  - with ExecutionRequest when Nodes from v2 engine are called by a workunit;
         #  - when a v2 engine Node is called by another v2 engine Node.
         self.parent_id = zipkin_attrs.span_id
 
-      span = zipkin_span(
+      span = local_tracer.zipkin_span(
         service_name=service_name,
         span_name=workunit.name,
         transport_handler=self.handler,
         encoding=Encoding.V1_THRIFT,
         zipkin_attrs=zipkin_attrs,
-        span_storage=self.span_storage,
       )
     else:
-      span = zipkin_span(
+      # If start_workunit is called from a new thread local_tracer doesn't have zipkin attributes.
+      # Parent's attributes need to be added to the local_tracer zipkin_attrs storage.
+      if not local_tracer.get_zipkin_attrs():
+        parent_attrs = workunit.parent.zipkin_attrs
+        local_tracer.push_zipkin_attrs(parent_attrs)
+        local_tracer.set_transport_configured(configured=True)
+      span = local_tracer.zipkin_span(
         service_name=service_name,
         span_name=workunit.name,
-        span_storage=self.span_storage,
       )
-    self._workunits_to_spans[workunit] = span
+    # For all spans except the first span zipkin_attrs for span are created at this point
     span.start()
+    if workunit.name == 'background' and workunit.parent == None:
+      span.zipkin_attrs = ZipkinAttrs(
+        trace_id=span.zipkin_attrs.trace_id,
+        span_id=span.zipkin_attrs.span_id,
+        parent_span_id=self.run_tracker._main_root_workunit.zipkin_attrs.span_id,
```

Comment:
Can you just set `span.zipkin_attrs.parent_span_id=self.run_tracker._main_root_workunit.zipkin_attrs.span_id` here instead of making a whole new `ZipkinAttrs`? Or does the constructor do more things?